bbot/modules/report/asn.py

```diff
@@ -38,25 +38,55 @@ async def filter_event(self, event):
 
     async def handle_event(self, event):
         host = event.host
-        if self.cache_get(host) is False:
-            asns, source = await self.get_asn(host)
-            if not asns:
-                self.cache_put(self.unknown_asn)
-            else:
-                for asn in asns:
-                    emails = asn.pop("emails", [])
-                    self.cache_put(asn)
-                    asn_event = self.make_event(asn, "ASN", parent=event)
-                    asn_number = asn.get("asn", "")
-                    asn_desc = asn.get("description", "")
-                    asn_name = asn.get("name", "")
-                    asn_subnet = asn.get("subnet", "")
-                    if not asn_event:
-                        continue
+        host_str = str(host)
+
+        # Check if this IP is already covered by a subnet we've processed
+        try:
+            ip_obj = ipaddress.ip_address(host_str)
+            for subnet_str, asn_number in self.processed_subnets.items():
+                try:
+                    subnet = ipaddress.ip_network(subnet_str, strict=False)
+                    if ip_obj in subnet:
+                        self.debug(
+                            f"IP {host_str} already covered by processed subnet {subnet_str} (ASN {asn_number})"
+                        )
+                        return
+                except ValueError:
+                    continue
+        except ValueError:
+            pass  # Invalid IP address, continue with normal processing
+
+        asn_data = await self.helpers.asn.ip_to_subnets(host_str)
+        if asn_data:
+            asn_record = asn_data
+            asn_number = asn_record.get("asn")
+            asn_desc = asn_record.get("description", "")
+            asn_name = asn_record.get("name", "")
+            asn_country = asn_record.get("country", "")
+            subnets = asn_record.get("subnets", [])
+
+            # Store ASN data locally for reporting
+            if asn_number and asn_number != "UNKNOWN" and asn_number not in self.asn_data:
+                self.asn_data[asn_number] = {
+                    "name": asn_name,
+                    "description": asn_desc,
+                    "country": asn_country,
+                    "subnets": set(subnets),
+                }
+                # Track processed subnets for quick lookups
```

Comment:
Since we're already storing a mapping of ASN ID -> ASN in the helper, we can rely on that.

For the module, it might make sense instead to keep a simple tally of ASN ID -> count of IPs per that ASN. Then in `.report()`, grab the ASN details using the helper.

Reply:
fixed in 5cd05d0e4